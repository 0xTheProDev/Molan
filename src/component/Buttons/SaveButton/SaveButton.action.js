// @flow
import { saveAs } from 'save-as';
import { MIME_TYPES, FILE_EXTENSIONS } from 'constants/templates';

export function onSaveButtonClick(language: string, content: string) {
  return function onClickHandler(event: MouseEvent): void {
    const MimeType = MIME_TYPES[language] || MIME_TYPES['text'];
    const Extension = FILE_EXTENSIONS[language] || FILE_EXTENSIONS['text'];

    const blob = new Blob([ content ], { type: MimeType });
    const FileName = `molan${Extension}`;

    saveAs(blob, FileName);
  }
}
