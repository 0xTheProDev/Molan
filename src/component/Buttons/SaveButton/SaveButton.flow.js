// @flow
export type SaveButtonProps = {
  language: string,
  content: string,
  SaveButtonText: string,
};

export type SaveButtonInternalProps = {
  SaveButtonText: string,
  onSaveButtonClick: (event: MouseEvent) => void,
};
