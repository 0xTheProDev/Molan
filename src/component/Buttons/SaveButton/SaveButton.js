// @flow
import React from 'react';
import { Button, Icon } from 'semantic-ui-react';
import { noop } from 'util/common';

import { SaveButtonInternalProps } from './SaveButton.flow';

export function SaveButton({ SaveButtonText, onSaveButtonClick }: SaveButtonInternalProps) {
  return (
    <Button
      animated="fade"
      floated="right"
      title={SaveButtonText}
      onClick={onSaveButtonClick}
    >
      <Button.Content hidden>{SaveButtonText}</Button.Content>
      <Button.Content visible>
        <Icon name="save" />
      </Button.Content>
    </Button>
  );
}

SaveButton.defaultProps = {
  SaveButtonText: 'Save',
  onSaveButtonClick: noop,
};
