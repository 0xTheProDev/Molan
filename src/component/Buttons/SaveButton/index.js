// @flow
import React from 'react';

import { SaveButtonProps } from './SaveButton.flow';
import { onSaveButtonClick } from './SaveButton.action';
import { SaveButton } from './SaveButton';

export default function SaveButtonFactory(props: SaveButtonProps) {
  const {
    language,
    content,
    SaveButtonText,
  } = props;

  const onClickHandler = onSaveButtonClick(props.language, props.content);

  return (
    <SaveButton
      SaveButtonText={SaveButtonText}
      onSaveButtonClick={onClickHandler}
    />
  );
}

export { SaveButtonProps, SaveButtonFactory as SaveButton };
