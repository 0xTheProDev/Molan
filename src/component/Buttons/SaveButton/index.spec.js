import React from 'react';
import { Button } from 'semantic-ui-react';
import * as SaveAs from 'save-as';

import SaveButton from './index';

describe('Component::Buttons:SaveButton::Integration Suite', () => {
  const
    Language   = 'c',
    Content    = '#include <stdio.h>\n\nint main(void) {\n\nreturn 0;\n}\n';

  let
    originalURLConstructor,
    saveAsStub,
    minProps;

  beforeAll(() => {
    saveAsStub = jest.spyOn(SaveAs, 'saveAs');

    originalURLConstructor = URL;
    global.URL = {
      createObjectURL: jest.fn(),
    };

    minProps = {
      SaveButtonText: 'Save',
      language: Language,
      content: Content,
    };
  });

  afterAll(() => {
    jest.restoreAllMocks();
    global.URL = originalURLConstructor;
  });

  it('should render with base structure', () => {
    const wrapper = mount(<SaveButton {...minProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should store content in file when user clicks on Save button', () => {
    const wrapper = mount(<SaveButton {...minProps} />);

    /** User clicks on Save Button */
    wrapper.find(Button).simulate('click');

    /** File should be prepared for download */
    expect(saveAsStub).toHaveBeenCalled();
  });
});
