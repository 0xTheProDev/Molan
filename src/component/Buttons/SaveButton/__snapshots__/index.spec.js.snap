// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component::Buttons:SaveButton::Integration Suite should render with base structure 1`] = `
<SaveButtonFactory
  SaveButtonText="Save"
  content="#include <stdio.h>

int main(void) {

return 0;
}
"
  language="c"
>
  <SaveButton
    SaveButtonText="Save"
    onSaveButtonClick={[Function]}
  >
    <Button
      animated="fade"
      as="button"
      floated="right"
      onClick={[Function]}
      title="Save"
    >
      <Ref
        innerRef={
          Object {
            "current": <button
              class="ui fade animated right floated button"
              title="Save"
            >
              <div
                class="hidden content"
              >
                Save
              </div>
              <div
                class="visible content"
              >
                <i
                  aria-hidden="true"
                  class="save icon"
                />
              </div>
            </button>,
          }
        }
      >
        <RefFindNode
          innerRef={
            Object {
              "current": <button
                class="ui fade animated right floated button"
                title="Save"
              >
                <div
                  class="hidden content"
                >
                  Save
                </div>
                <div
                  class="visible content"
                >
                  <i
                    aria-hidden="true"
                    class="save icon"
                  />
                </div>
              </button>,
            }
          }
        >
          <button
            className="ui fade animated right floated button"
            onClick={[Function]}
            title="Save"
          >
            <ButtonContent
              hidden={true}
            >
              <div
                className="hidden content"
              >
                Save
              </div>
            </ButtonContent>
            <ButtonContent
              visible={true}
            >
              <div
                className="visible content"
              >
                <Icon
                  as="i"
                  name="save"
                >
                  <i
                    aria-hidden="true"
                    className="save icon"
                    onClick={[Function]}
                  />
                </Icon>
              </div>
            </ButtonContent>
          </button>
        </RefFindNode>
      </Ref>
    </Button>
  </SaveButton>
</SaveButtonFactory>
`;
